multidim-interop/rust/src/lib.rs

```diff
@@ -78,7 +78,9 @@ where
     let mut conn = client.get_async_connection().await?;
 
     info!("Running ping test: {}", swarm.local_peer_id());
-    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
+    env_logger::Builder::from_env(Env::default().default_filter_or("info"))
+        .target(Target::Stdout)
```

Comment:
AFAIK, it is good unix behaviour to send logs to stderr. Why change to stdout?

Reply:
So as they show on the current interop tests, currently interop tests only show `stdout` see [here](https://github.com/libp2p/test-plans/blob/master/multidim-interop/src/compose-runner.ts#L61).  
I read a couple of discussions like [this one](https://softwareengineering.stackexchange.com/questions/439462/log-levels-and-stdout-vs-stderr) as I also found it surprising that the default for `env_logger` was `stderr` but as I didn't see an industry wide statys quo, so I thought it would be easier to just change the Rust tests, since the Go ones already log to `stdout`.

Reply:
at least the “ping successful” should be on stdout. The rest can be on stderr. https://pubs.opengroup.org/onlinepubs/9699919799/functions/stderr.html

Reply:
If `ping successful` is part of the test contract, we should probably be printing that on `stdout` with `println`. Logs should be on `stderr` IMO.

I am sure you thought about this @MarcoPolo: It would be nice if we could rely on the exit code instead of scanning `stdout` for a string. Would it make sense to push a `listenerDone` to redis and thus synchronise the shutdown between the two?

Reply:
This is untested but have you explored using `depends-on` to express the relationship between `dialer` and `listener`? If we tell `docker compose` that `dialer` depends on `listener` and have `listener` run an endless loop at the end, we may get a cleaner shutdown behaviour where only the dialer exits by itself.

Reply:
> If ping successful is part of the test contract, we should probably be printing that on stdout with println. Logs should be on stderr IMO.

Agreed on principle.

> I am sure you thought about this @MarcoPolo: It would be nice if we could rely on the exit code instead of scanning stdout for a string. Would it make sense to push a listenerDone to redis and thus synchronise the shutdown between the two?

Yes. That's what happens. docker compose runs `up` with `--exit-code-from dialer` and uses that exit code as the result of the test. The unfortunate thing is that option implies `--abort-on-container-exit` which will stop all containers if any container exits. If the listener exits first, then we'll get a non-zero exit code from dialer because it was stopped by compose. Extra synchronization doesn't fix this because the dialer can still be slow in shutting down. This lead to some flakiness in the js-libp2p tests.

We may be able to start these in a detached mode, then wait for the exit of the dialer, then stop the compose instances. A bit more involved but doable.

---

The reason I want the `Ping Successful: <duration>` log output is because then these tests will just work with the round trip counter with no modifications. We would introduce latency in the network link, then parse the duration.

Reply:
> have listener run an endless loop at the end, we may get a cleaner shutdown behaviour where only the dialer exits by itself.

That's also an alternative. Although at first glance, I'm not a fan of the wait endlessly for our process to be terminated.

Reply:
> > have listener run an endless loop at the end, we may get a cleaner shutdown behaviour where only the dialer exits by itself.
> 
> That's also an alternative. Although at first glance, I'm not a fan of the wait endlessly for our process to be terminated.

Would likely be wise to have a timeout around that.

Reply:
> We may be able to start these in a detached mode, then wait for the exit of the dialer, then stop the compose instances. A bit more involved but doable.

Ugh, I can definitely see the appeal in having docker compose manage all of that for us.

> The reason I want the `Ping Successful: <duration>` log output is because then these tests will just work with the round trip counter with no modifications. We would introduce latency in the network link, then parse the duration.

Can you expand on what you mean by "the round trip counter"?

I am not opposed to logging success and failure of the test and parse further results but I'd prefer if we can do that conditionally on the exit code. In other words: Have the exit code be the source of truth for test success / failure and then parse stdout/stderr afterwards if we wanted to get more information. We might want to consider emitting JSON then so we can more easily parse it from our test runner.

Reply:
> Can you expand on what you mean by "the round trip counter"?

I assume that would be https://github.com/libp2p/test-plans/issues/52, but maybe the cleaner solution is to output a file, since we probably want to export multiple values?

Reply:
What multiple values?

thinking about it more I like the “leave listener around for a bit”. It removes a redis sync point and lets us use the exit code as source of truth.

Reply:
It would be a little easier to parse JSON. But I don't have a preference here. Matching against output strings is only marginally more work. And it's a little nicer to see "Ping Successful: <duration>" then `{"pingDuration": <duration>}`. But this is purely subjective.

Happy to define the stdout schema to print a JSON result at the end.

Reply:
> What multiple values?

1. RTT, as measured by the ping protocol
2. Duration of the handshake

Reply:
> What multiple values?
> 
> thinking about it more I like the “leave listener around for a bit”. It removes a redis sync point and lets us use the exit code as source of truth.

I opened an issue: https://github.com/libp2p/test-plans/issues/115

Reply:
> And it's a little nicer to see "Ping Successful: " then `{"pingDuration": <duration>}`.

Depending on how we define the JSON structure, the test runner can parse it and print something more human-centric :)

Reply:
Following up on the concrete discussion for this pull request I suggest:

- Logging to stderr
- Print `ping successful` to stdout via `println!`.